src/coap.c

```diff
@@ -308,4 +308,180 @@ NS_INTERNAL uint32_t coap_parse(struct iobuf *io, struct ns_coap_message *cm) {
+  }
+
+  return ret;
+}
+
+/* Splits given Opt Number/Length into base and ext values. */
+static int coap_split_opt(uint32_t value, uint8_t *base, uint16_t *ext) {
+  int ret = 0;
+
+  if (value < 13) {
+    *base = value;
+  } else
+  if (value >= 13 && value <= 0xFF + 13) {
+    *base = 13;
+    *ext = value - 13;
+    ret = sizeof(uint8_t);
+  } else if (value > 0xFF + 13 && value <= 0xFFFF + 269) {
+    *base = 14;
+    *ext = value - 269;
+    ret = sizeof(uint16_t);
+  }
+
+  return ret;
+}
+
+/* Puts uint16_t (in network order) into given char stream. */
+static char *coap_add_uint16(char *ptr, uint16_t val) {
+  *ptr = val >> 8;
+  ptr++;
+  *ptr = val & 0x00FF;
+  ptr++;
+  return ptr;
+}
+
+/* Puts extended value of Opt Number/Length into given char stream. */
+static char *coap_add_opt_info(char *ptr, uint16_t val, size_t len) {
+  if (len == sizeof(uint8_t)) {
+    *ptr = val;
+    ptr++;
+  } else if (len == sizeof(uint16_t)) {
+    ptr = coap_add_uint16(ptr, val);
+  }
+
+  return ptr;
+}
+
+/* Verifies given cm and calculates message size for it. */
+static uint32_t coap_calculate_packet_size(struct ns_coap_message *cm,
+                                           size_t *len) {
+  struct ns_coap_option *opt;
+  uint32_t prev_opt_number;
+
+  *len = 4;  /* header */
+  if (cm->msg_type > NS_COAP_MSG_MAX) {
+    return NS_COAP_ERROR | NS_COAP_MSG_TYPE_FIELD;
+  }
+  if (cm->token.len > 8) {
+    return NS_COAP_ERROR | NS_COAP_TOKEN_FIELD;
+  }
+  if (cm->code_class > 7 ) {
+    return NS_COAP_ERROR | NS_COAP_CODE_CLASS_FIELD;
+  }
+  if (cm->code_detail > 31) {
+    return NS_COAP_ERROR | NS_COAP_CODE_DETAIL_FIELD;
+  }
+
+  *len += cm->token.len;
+  if (cm->payload.len != 0) {
+    *len += cm->payload.len + 1;  /* ... + 1; add payload marker */
+  }
+
+  opt = cm->options;
+  prev_opt_number = 0;
+  while (opt != NULL) {
+    *len += 1;  /* basic delta/length */
+    *len += coap_get_ext_opt_size(opt->number);
+    *len += coap_get_ext_opt_size((uint32_t)opt->value.len);
+    /* 
+     * Current implementation performs check if
+     * option_number > previous option_number and produces an error
+     * TODO(alashkin): write design doc with limitations
+     * May be resorting is more suitable solution.
+     */
+    if ((opt->next != NULL && opt->number > opt->next->number)
+        || opt->value.len > 0xFFFF + 269
+        || opt->number - prev_opt_number > 0xFFFF + 269) {
+      return NS_COAP_ERROR | NS_COAP_OPTIONS_FIELD;
+    }
+    *len += opt->value.len;
+    opt = opt->next;
+  }
+
+  return 0;
+}
+
+/*
+ * Composes CoAP message from cm structure.
+ * Returns 0 on success or bit mask with wrong field.
+ * Packet returned must be explicitly deleted by NS_FREE.
+ */
+NS_INTERNAL uint32_t coap_compose(struct ns_coap_message *cm,
+                                  char **packet, size_t *len) {
+  struct ns_coap_option *opt;
+  uint32_t res, prev_opt_number;
+  char *ptr;
+
+  res = coap_calculate_packet_size(cm, len);
```

Comment:
That document is about another optimisation, dealing with consuming iobufs.

Preallocating is definitely better. You can preallocate via the iobuf interface as well, so you can add the code that does this optimisation later. Some boards have not much code memory as well, so perhaps that's the trade off one wants to make (e.g. compile out the part that preallocates and pay a footprint or cpu penalty).

On the other hand reallocs (at least) double the memory usage during realloc, which arguably is not so important in single threaded environments, unless you're really walking on the edge. Furthermore we can always implement a sparse buffer solution. The rationale for the single contiguous iobuf is mostly to simplify parsing.

I'm not sure what's the best approach here, especially given that CoAP is especially targeted at extremely low memory environments.

I'd be tempted to suggest a first draft with simple implementation that fits well in fossa and tweak it later.